### ISSUE TITLE
Rule proposal: Prefer memoized callback functions

### ISSUE THREAD

**Issue by oliveryasuna** [jsx-eslint/eslint-plugin-react#3683]
Consider the following:

```typescript jsx
// Option 1.
const MyComponent = () => {
  return <div onClick={() => console.log('clicked')}>Click me</div>;
};

// Option 2.
const MyComponent = () => {
  const handleClick = () => {
    console.log('clicked');
  };
  
  return <div onClick={handleClick}>Click me</div>;
};
```

Both options behave the same. The callback is recreated on every re-render. One might prefer to memoized the callback to avoid this common pitfall:

```typescript jsx
// Option 3.
const MyComponent = () => {
  const handleClick = useCallback(() => {
    console.log('clicked');
  }, []);

  return <div onClick={handleClick}>Click me</div>;
};
```

I propose a new rule that reports when a callback function is defined in JSX or would be recreated on every re-render. More generally, such a rule could allow developers to prefer one pattern over the other.

_Note: I opened this issue is conjunction with https://github.com/jsx-eslint/eslint-plugin-react/issues/3682._
